### ISSUE TITLE
Type "{ a: false | true }" is incompatible with type "{ a: false } | { a: true }"

### ISSUE THREAD

**Issue by thecotne** [JSMonk/hegel#181]
```js
type Data = { a: true } | { a: false }

function thisDoesNot(data: Data): Data {
  return {
    a: data.a
  }
}
```

[try](https://hegel.js.org/try#C4TwDgpgBAIghsOUC8UDeU4C4rAE4Cu0AvlAD7qY4BmcANgM4kBQz1BAdgMbACWA9h1wALXgxj8IDAHL9gACgAmCbLBUBKHPETpmUKHgjACeIWj36qUZYgB0cC8WbEgA)

**Comment by thecotne** [comment]
message has changed from `Type "{ a: false | true }" is incompatible with type "{ a: false } | { a: true }"` to `Type "{ a: boolean }" is incompatible with type "{ a: false } | { a: true }"`

i think change was introduced with this commit https://github.com/JSMonk/hegel/commit/95fdae84819b1f131110456f7e1a1f7aa2d39da1

**Comment by JSMonk** [comment]
Yes. And I don't know how to fix it for now :( I'm about the issue.
The issue exists in Flow [Example](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgAiAhhsWALxgDeYxAXGBgE4CuBAvmAD412NRiMAM6d0UVgDsAxhgCWcSUwAWc4YTh5hAOTgYAFABNSDIiYCUjEmRqowYZngytmS6nfv8wxsgDpiHhyoHEA).
And doesn't exist in TypeScript [Example](https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=8&pc=25#code/C4TwDgpgBAIghsOUC8UDeU4C4rAE4Cu0AvlAD7qY4BmcANgM4kBQz1BAdgMbACWA9h1wALXgxj8IDAHL9gACgAmCbLBUBKHPETpmUKHgjACeIWj36qUZYgB0cC8WbEgA).
But with TypeScript, there is problem, because of type `{ a: false | true }` equal to type `{ a: false } | { a: true }`. And we easily can get the next error:
```typescript
function process(a: { value: true } | { value: false }) {
  a.value = false;
}

const CONSTANT: { value: true } = { value: true };

process(CONSTANT);
```

[Try](https://www.typescriptlang.org/play/index.html?ssl=5&ssc=1&pln=1&pc=1#code/GYVwdgxgLglg9mABABwE5wgUwM7YBQCGAXIgN6IBuBANiJiVKnYgL6IA+ZlNdJwN2TKwCUZAFCJEBAHRVaQgLyJ+1QQG4xLMWIgJsURAGEA8gDkAygBUAgqcslyc3okbM2Sxz3oumQlhrE0DBx8EwsbO2FtIA)

**Comment by thecotne** [comment]
@JSMonk that's how typescript's type assignment works there is no safety or soundness in it see https://github.com/microsoft/TypeScript/issues/33748

but i understand why it's complicated to implement 

i think this should work as well (does not work in ts and flow)

```ts
type Data = { a: string } | { a: number }

function thisDoesNot(data: Data): Data {
  return {
    a: data.a
  }
}
```

[try](https://hegel.js.org/try#C4TwDgpgBAIghsOUC8UDeU4C4oGdgBOAlgHYDmUAvlAD7qY4kCuAtgEYQFUBQ3AZkxIBjYEQD2JKMAAWRXDDERcAOTHAAFABME2WDoCUOeInTcoUAhGBMCktGfMMo2xADo4Dyt0pA)


editing of union object is different story for that you need to refine type to sub set of union objects  that allow type of modification developer is intended to do for example


```ts
type Data = { a: string } | { a: number }

function notOk(data: Data) {
  data.a = 'sdsds'//err
}

function ok(data: Data) {
  if (typeof data.a === 'string') {
    data.a = 'sds'//ok
  }
  if (typeof data.a === 'number') {
    data.a = 111//ok
  }
}
```
